File: gget-java/src/main/java/cz/zakjan/gget/visitor/GetByPathVisitorFunctions.java
package cz.zakjan.gget.visitor;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class GetByPathVisitorFunctions {
    private static Map<String, Function<List<Object>, Object>> nameToFunction = new HashMap<String, Function<List<Object>, Object>>() {{
        put("join", GetByPathVisitorFunctions::joinFunction);
        put("split", GetByPathVisitorFunctions::splitFunction);
        put("sum", GetByPathVisitorFunctions::sumFunction);
    }};

    public static Object callFunction(String name, List<Object> args) {
        Function<List<Object>, Object> func = nameToFunction.get(name);
        if (func == null) {
            return null;
        }

        return func.apply(args);
    }

    private static Object joinFunction(List<Object> args) {
        if (args.size() != 2) {
            return null;
        }

        Object values = args.get(0);
        Object delimiter = args.get(1);
        if (!(values instanceof List && delimiter instanceof String)) {
            return null;
        }

        List<String> filteredValues = ((List<Object>)values).stream().filter(x -> x instanceof String).map(x -> (String)x).collect(Collectors.toList());
        if (filteredValues.size() == 0) {
            return null;
        }

        String result = String.join((String)delimiter, filteredValues);
        return result;
    }

    private static Object splitFunction(List<Object> args) {
        if (args.size() != 2) {
            return null;
        }

        Object value = args.get(0);
        Object delimiter = args.get(1);
        if (!(value instanceof String && delimiter instanceof String)) {
            return null;
        }

        List<String> result = Arrays.asList(((String)value).split((String)delimiter));
        return result;
    }

    private static Object sumFunction(List<Object> args) {
        if (args.size() != 1) {
            return null;
        }

        Object values = args.get(0);
        if (!(values instanceof List)) {
            return null;
        }

        List<Integer> filteredValues = ((List<Object>)values).stream().filter(x -> x instanceof Integer).map(x -> (Integer)x).collect(Collectors.toList());
        if (filteredValues.size() == 0) {
            return null;
        }

        Integer result = filteredValues.stream().reduce(0, (sum, x) -> sum + x);
        return result;
    }
}

File: gget-ts/src/visitor/GetByPathVisitorFunctions.ts
export class GetByPathVisitorFunctions {
    private static nameToFunction = new Map([
        ["join", GetByPathVisitorFunctions.joinFunction],
        ["split", GetByPathVisitorFunctions.splitFunction],
        ["sum", GetByPathVisitorFunctions.sumFunction],
        ["timestampToIsoString", GetByPathVisitorFunctions.timestampToIsoStringFunction],
        ["isoStringToTimestamp", GetByPathVisitorFunctions.isoStringToTimestampFunction],
    ]);

    public static callFunction(name: string, args: unknown[]): unknown {
        const func = GetByPathVisitorFunctions.nameToFunction.get(name);
        /* istanbul ignore if */
        if (!func) {
            return null;
        }

        return func(args);
    }

    private static joinFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 2) {
            return null;
        }

        const values = args[0];
        const delimiter = args[1];
        /* istanbul ignore if */
        if (!(Array.isArray(values) && typeof delimiter === "string")) {
            return null;
        }

        const filteredValues = values.filter(x => typeof x === "string");
        /* istanbul ignore if */
        if (filteredValues.length === 0) {
            return null;
        }

        const result = filteredValues.join(delimiter)
        return result;
    }

    private static splitFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 2) {
            return null;
        }

        const value = args[0];
        const delimiter = args[1];
        /* istanbul ignore if */
        if (!(typeof value === "string" && typeof delimiter === "string")) {
            return null;
        }

        const result = value.split(delimiter);
        return result;
    }

    private static sumFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 1) {
            return null;
        }

        const values = args[0];
        /* istanbul ignore if */
        if (!Array.isArray(values)) {
            return null;
        }

        const filteredValues = values.filter(x => typeof x === "number");
        /* istanbul ignore if */
        if (filteredValues.length === 0) {
            return null;
        }

        const result = filteredValues.reduce((sum, x) => sum + x, 0);
        return result;
    }

    private static timestampToIsoStringFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 1) {
            return null;
        }

        const value = args[0];
        /* istanbul ignore if */
        if (typeof value !== "number") {
            return null;
        }

        return new Date(value).toISOString();
    }

    private static isoStringToTimestampFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 1) {
            return null;
        }

        const value = args[0];
        /* istanbul ignore if */
        if (typeof value !== "string") {
            return null;
        }

        return Date.parse(value).valueOf();
    }
}